calculators/gas_king_rating.php

```diff
@@ -0,0 +1,122 @@
+    {
+        foreach($team as &$player)
+        {
+            $diff = 0;
+
+            if ($player['status'] == 'win')
+            {
+                for ($i = 0; $i < $player['gas_count']; $i++)
+                {
+                    $gas_timing_in_minutes = intval($player['gas_timings'][$i]) / $number_of_frames_in_a_min; 
+                    $diff -= 10 * ($gas_timing_in_minutes - $min_timings[$i]);
+                    $diff += ($i + 1) * ($duration_of_the_game_in_minutes - $gas_timing_in_minutes);
+                }
+            }
+
+            else
+            {
+                if($player['gas_count'] == 0)
+                {
+                    $diff -= 4 * 10 * $duration_of_the_game_in_minutes;
+                }
+
+                else
+                {
+                    for ($i = 0; $i <= intval($player['gas_count']); $i++)
+                    {
+                        $gas_timing_in_minutes = intval($player['gas_timings'][$i]) / $number_of_frames_in_a_min;
+
+                        if ($i > 0)
+                        {
+                            $punishing_for_gassing_too_soon = 10 * ($gas_timing_in_minutes - (intval($player['gas_timings'][$i - 1]) / $number_of_frames_in_a_min + $good_timings[$i - 1]));
+
+                            if ($punishing_for_gassing_too_soon < 0)
+                            {
+                                $diff += $punishing_for_gassing_too_soon;
+                            }
+                        }
+
+                        $diff -= ($i + 1) * ($duration_of_the_game_in_minutes - $gas_timing_in_minutes);
+                    }
+                }
+            }
+
+            if ($player['rating'] + $diff >= 0)
+            {
+                $dist_to_2000 = $player['rating'] - 2000;
+
+                if ($dist_to_2000 >= 0)
+                {
+                    if (dsr_gas_king_rating__same_sign($dist_to_2000, $diff))
+                    {
+                        $diff *= exp(-($dist_to_2000) / 2000);
+                    }
+
+                    else
+                    {
+                        if ($diff + $dist_to_2000 < 0)
+                        {
+                            $diff = ($diff + $dist_to_2000) / 10 - $dist_to_2000;
+                        }
+                    }
+                }
+
+                else
+                {
+                    if (dsr_gas_king_rating__same_sign($dist_to_2000, $diff))
+                    {
+                        $diff /= 10;
+                    }
+
+                    else
+                    {
+                        if ($diff / 10 + $dist_to_2000 > 0)
+                        {
+                            $diff += $dist_to_2000 * 10 - $dist_to_2000;
+                        }
+
+                        else
+                        {
+                            $diff /= 10;
+                        }
+                    }
+                }
+
+                $player['rating'] += round($diff);
```

Comment:
your diffs can be negative, right? Then it is on you to assure rating will never go below 0, or 1, or whatever positive number you want to have as lowest number.
Like this: `$player['rating'] = max(1, $player['rating'] + $diff);`

Since you've set the starting rating to 2000, lowest should probably be 1, not 0.

Reply:
also refer to the manual for the functions you are using, https://www.php.net/manual/en/function.round.php , it says round() returns float. Rating must contain int. Do it `$player['rating'] = max(1, $player['rating'] + intval($diff));`

Reply:
Replaced the round by intval, I will check the diff problem tomorrow

Reply:
I checked the diff, it couldn't go in the negatives but could reach 0, now it can't anymore, fixed